spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/QueryMapper.java

```diff
@@ -862,8 +862,9 @@ public MongoPersistentProperty getProperty() {
 		 */
 		@Override
 		public MongoPersistentEntity<?> getPropertyEntity() {
+
 			MongoPersistentProperty property = getProperty();
-			return property == null ? null : mappingContext.getPersistentEntity(property);
+			return property != null && property.isEntity() ? mappingContext.getPersistentEntity(property) : null;
```

Comment:
I wonder if we should rather make the `….isEntity()` check part of the implementation of `getPersistentEntity(PersistentProperty)`. That would prevent that any other calls to that method with a property that's not an entity in the first place would trigger the addition of the type to the mapping context and the code here could even stay untouched.

Reply:
Changing either one is rather a patch than a fix to the actual issue. There's no coherent behavior between `PersistentProperty.isEntity()` and `AbstractMappingContext.shouldCreatePersistentEntityFor(…)` as both run different checks.

Reply:
I generally agree, however, I'd argue it sort of has to be that way (to support the use case you described above). What I am trying to say is, on `getPersistentEntity(PersistentProperty)`, we *know* that we're in the "nested" case, so that stricter rules might apply and actually can be applied. I.e. in case of the lookup of the `PersistentEntity` for a `PersistentProperty` we can apply more strict rules before adding a previously not-known type to the `MappingContext` than when asking for a `PersistentEntity` on the `MappingContext` as that implicitly assumes top-level types to some degree.

I'd also it's inline with the previously implicit assumption that looking up a `PersistentEntity` for a property, that the latter actually is considered one.

Reply:
You're right about the relationship between a property and its reference to a top-level type. We generally want to avoid introspection of types that are considered simple ones.

Let's do it that way. I'll file a separate ticket to address the missing simple type check in `shouldCreatePersistentEntityFor(…)`.